### PR TITLE
Add @OptionalChild property wrapper (one-to-one relation)

### PR DESCRIPTION
**Pull request by fananek**
Add new `@OptionalChild` property for one-to-one relations.

Example:
```swift
private final class Planet: Model {
    static let schema = "planets"

    @ID(key: .id)
    var id: UUID?

    @Field(key: "name")
    var name: String

    @OptionalChild(for: \.$planet)
    var governor: Governor

    init() { }

    init(id: IDValue? = nil, name: String) {
        self.id = id
        self.name = name
    }
}

private final class Governor: Model {
    static let schema = "governors"

    @ID(key: .id)
    var id: UUID?

    @Field(key: "name")
    var name: String

    @Parent(key: "planet_id")
    var planet: Planet

    init() { }

    init(id: IDValue? = nil, name: String, planetID: UUID) {
        self.id = id
        self.name = name
        self.$planet.id = planetID
    }
}
```

See test files `ChildTests.swift` and `Governor.swift` for more details.

This PR:
- Closes issue #74 
- Supersedes #394 (which seems to be inactive)

### CODE REVIEW

File: Sources/FluentKit/Properties/Child.swift [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        to.forEach {
+            switch self.parentKey {
+            case .required(let keyPath):
+                $0[keyPath: keyPath].id = id
+            case .optional(let keyPath):
+                $0[keyPath: keyPath].id = id
+            }
+        }
+        return to.create(on: database)
+    }
```
Comment (siemensikkema) [inline Sources/FluentKit/Properties/Child.swift:77]: This method kind of goes against the idea of there being a single child.

**Review by siemensikkema — COMMENTED** [review]
Thanks for your PR! Am I correct in understanding that this in essence offers the same functionality as doing `.first` on `Children` property? This is not to downplay your work but just for my understanding. As in, it does not seem to offer any guarantees that are not more `Child`ren for a given parent model. That would still need to be enforced through a database constraint for instance.

File: Sources/FluentKit/Properties/Child.swift (hunk 1 above)
Comment (fananek) [reply Sources/FluentKit/Properties/Child.swift]: Thanks for reviewing the code. You are absolutely right in both points.
- The method is a leftover I missed. I already pushed new commit to PR.
- Current implementation requires `.unique` database constraint to ensure `Child` uniqueness. I'm open to any ideas if there is a feasible way to check whether `Child` exists during creation.

**Comment by 0xTim** [comment]
This should allow you to define a single child property on the parent `var child: MyChild` rather than an array. This definitely simplifies several code paths instead of having to deal with an array

**Comment by fananek** [comment]
@siemensikkema. @0xTim - Hi guys, is there anything else I can do in order to successfully merge and close this PR?

File: Sources/FluentKit/Properties/Child.swift [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            self.idValue = try output.decode(key, as: From.IDValue.self)
+        }
+    }
+}
+
+// MARK: Codable
+
+extension ChildProperty: AnyCodableProperty {
+    public func encode(to encoder: Encoder) throws {
+        if let rows = self.value {
```
Comment (0xTim) [inline Sources/FluentKit/Properties/Child.swift:116]: This should probably be a single row or renamed to `child` since it's no longer an array

**Comment by 0xTim** [comment]
This is great! Thanks for this. My only query is whether `@Child` should only be applied to 
optional properties. When you first create a `@Parent` there's no child so getting the model and eager loading may cause issues. Additionally, I'd like to see a test for `query(on:)` for Child just to show how it works - I'm assuming this requires a `.first()` to get the actual value?

**Comment by 0xTim** [comment]
I'm happy to defer to a lot of the community for this one though, but I'd like @gwynne to chime in as well

File: Sources/FluentKit/Properties/Child.swift (hunk 2 above)
Comment (fananek) [reply Sources/FluentKit/Properties/Child.swift]: Good catch, thanks. New commit already in place.

**Comment by 0xTim** [comment]
Just to follow up from Discord - the consensus is that `@Child` should be restricted to optional values only

**Comment by fananek** [comment]
@0xTim Hi Tim, the latest version is available for review. Hope I got it right.
- Adjusted Child implementation.
- `@Child` has been renamed to `@OptionalChild` so we keep naming consistent.
- There are few more `query(on: )` tests for different relations

File: Sources/FluentBenchmark/SolarSystem/Governor.swift [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        self.id = id
+        self.name = name
+        self.$planet.id = planetId
+    }
+}
+
+public struct GovernorMigration: Migration {
+    public func prepare(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("governors")
+            .field("id", .uuid, .identifier(auto: false))
```
Comment (gwynne) [inline Sources/FluentBenchmark/SolarSystem/Governor.swift:32]: Add `.unique` and `.required` constraints here for completeness' sake.
```suggestion
            .field(.id, .uuid, .identifier(auto: false), .required, .unique)
```

File: Sources/FluentBenchmark/SolarSystem/Governor.swift [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                    governor = .init(name: "John Doe")
+                case "Earth":
+                    governor = .init(name: "Jane Doe")
+                default:
+                    governor = nil
+                }
+                guard let gov = governor else {
+                    return database.eventLoop.makeSucceededFuture(())
+                }
+                return planet.$governor.create(gov, on: database)
```
Comment (gwynne) [inline Sources/FluentBenchmark/SolarSystem/Governor.swift:62]:
```suggestion
                default:
                    return database.eventLoop.future(())
                }
                return planet.$governor.create(governor!, on: database)
```

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    init(id: IDValue? = nil, name: String) {
+        self.id = id
+        self.name = name
+    }
+}
+
+private struct FooMigration: Migration {
+    func prepare(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("foos")
+            .field("id", .uuid, .identifier(auto: false))
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:96]:
```suggestion
            .field(.id, .uuid, .identifier(auto: false), .required, .unique)
```

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        self.id = id
+        self.bar = bar
+        self.$foo.id = fooID
+    }
+}
+
+private struct BarMigration: Migration {
+    func prepare(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("bars")
+            .field("id", .uuid, .identifier(auto: false))
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:130]:
```suggestion
            .field(.id, .uuid, .identifier(auto: false), .required, .unique)
```

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    init(id: IDValue? = nil, bar: Int, fooID: Foo.IDValue) {
+        self.id = id
+        self.bar = bar
+        self.$foo.id = fooID
+    }
+}
+
+private struct BarMigration: Migration {
+    func prepare(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("bars")
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:129]:
```suggestion
        database.schema(Bar.schema)
```

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+    init(id: IDValue? = nil, name: String) {
+        self.id = id
+        self.name = name
+    }
+}
+
+private struct FooMigration: Migration {
+    func prepare(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("foos")
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:95]:
```suggestion
        database.schema(Foo.schema)
```

File: Sources/FluentBenchmark/SolarSystem/Governor.swift [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        database.schema("governors")
+            .field("id", .uuid, .identifier(auto: false))
+            .field("name", .string, .required)
+            .field("planet_id", .uuid, .required, .references("planets", "id"))
+            .unique(on: "planet_id")
+            .create()
+    }
+
+    public func revert(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("governors").delete()
```
Comment (gwynne) [inline Sources/FluentBenchmark/SolarSystem/Governor.swift:40]:
```suggestion
        database.schema(Governor.schema).delete()
```

File: Sources/FluentBenchmark/SolarSystem/Governor.swift [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    public init(id: IDValue? = nil, name: String, planetId: UUID) {
+        self.id = id
+        self.name = name
+        self.$planet.id = planetId
+    }
+}
+
+public struct GovernorMigration: Migration {
+    public func prepare(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("governors")
```
Comment (gwynne) [inline Sources/FluentBenchmark/SolarSystem/Governor.swift:31]:
```suggestion
        database.schema(Governer.schema)
```

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            .field("id", .uuid, .identifier(auto: false))
+            .field("name", .string, .required)
+            .create()
+    }
+
+    func revert(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("foos").delete()
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:102]:
```suggestion
        database.schema(Foo.schema).delete()
```

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        database.schema("bars")
+            .field("id", .uuid, .identifier(auto: false))
+            .field("bar", .int, .required)
+            .field("foo_id", .uuid, .required)
+            .unique(on: "foo_id")
+            .create()
+    }
+
+    func revert(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("bars").delete()
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:138]:
```suggestion
        database.schema(Bar.schema).delete()
```

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        self.id = id
+        self.baz = baz
+        self.$foo.id = fooID
+    }
+}
+
+private struct BazMigration: Migration {
+    func prepare(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("bazs")
+            .field("id", .uuid, .identifier(auto: false))
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:166]:
```suggestion
        database.schema(Baz,schema)
            .field(.id, .uuid, .identifier(auto: false), .required, .unique)
```

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+            .field("baz", .double, .required)
+            .field("foo_id", .uuid)
+            .create()
+    }
+
+    func revert(on database: Database) -> EventLoopFuture<Void> {
+        database.schema("bazs").delete()
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:173]:
```suggestion
        database.schema(Baz.schema).delete()
```

**Review by gwynne — CHANGES_REQUESTED** [review]
Looks great, most of my comments are minor nits. One request: Can we add a SQL-mode-only test for ensuring `@OptionalChild` also works with `@ID` columns that are not `UUID`s? (SQL mode can be detected either by the mechanism in `SQLTests.swift` of checking `database as? SQLDatabase`, or by one of my and Tanner's several inconsistent attempts to provide it as an input to the benchmarks, as in `SortTests.swift` 🤦‍♀️)

I'll also see if I can't get another at least incremental update to the CI done.

**Comment by gwynne** [comment]
@fananek This is excellent work, thank you for doing this!

File: Sources/FluentBenchmark/Tests/ChildTests.swift [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+private final class Game: Model {
+    static let schema = "games"
+
+    @ID(custom: .id, generatedBy: .database)
+    var id: Int?
+
+    @Field(key: "title")
+    var title: String
+
+    // It's a solitare game :P
```
Comment (gwynne) [inline Sources/FluentBenchmark/Tests/ChildTests.swift:221]: I lol'd at this - well played! 🤣

**Review by gwynne — APPROVED** [review]
I love it! Cut, print it, ship it!! 💯💯

**Comment by gwynne** [comment]
@fananek One last item - the pull request's title and description are used by the bot to create the new release once the PR is merged; can you update the text accordingly? If you aren't sure what to change, I can take care of it, but I always try to give the original author first crack 🙂.

**Comment by fananek** [comment]
@gwynne, @0xTim, @siemensikkema  Big thanks to all participants for reviews, suggestions and overall support with this PR!

File: Sources/FluentBenchmark/Tests/ChildTests.swift (hunk 15 above)
Comment (fananek) [reply Sources/FluentBenchmark/Tests/ChildTests.swift:221]: Just a small `easter egg` comment. :)

**Comment by tanner0101** [comment]
These changes are now available in [1.11.0](https://github.com/vapor/fluent-kit/releases/tag/1.11.0)

**Pull request merged**